manager/integration/negative_tests/negative.robot

```diff
@@ -0,0 +1,24 @@
+*** Settings ***
+Documentation    Negative Test Cases
+
+Library    ./lib/Operations.py
+
+Variables    ./lib/Operations.py
+
+*** Variables ***
+@{NODE_IDS}=    ip-10-0-1-154    ip-10-0-1-74    ip-10-0-1-102
+${LOOP_COUNT}    200
+
+*** Test Cases ***
+Replica Rebuilding With Replica Deletion
+    Create Volume    volume-1
+    ${MOUNT_PATH}=    Format Volume    volume-1    -b 1024
+    Write Random Data To Volume    ${MOUNT_PATH}/data
+    ${CHECKSUM}=    Get Volume Data Checksum    ${MOUNT_PATH}/data
+    FOR    ${i}    IN    RANGE    ${LOOP_COUNT}
+        Delete Replica To Trigger Rebuilding    volume-1    ${NODE_IDS}[0]
+        Delete Replica To Trigger Rebuilding    volume-1    ${NODE_IDS}[1]
+        Delete Replica To Trigger Rebuilding    volume-1    ${NODE_IDS}[2]
+        ${CHECKSUM_PRIME}=    Get Volume Data Checksum    ${MOUNT_PATH}/data
+        Should Be Equal    ${CHECKSUM}    ${CHECKSUM_PRIME}
+    END
```

Comment:
For the upper layer of each test case, suggest making it human-readable, and the detailed steps can be implemented in separate keyword in the same robot file. Later, we can use the test case to generate the test cases for user reference.

For example here can be. Also, if keywords can be reused, you can place them in a separate dir to make them reusable for other test suites/cases.

```robot
*** Test Cases ***
Replica Rebuilding With Replica Deletion
    Given create Volume
    and write random data to volume
    When delete replica to trigger replica rebuilding
    Then after replica rebuilt, the volume data is intact
 
 *** Keywords ***
  Create Volume
  ..
  
  Write Random Data to Volume
  ...
```

BTW, this is not a negative test case but a normal replica rebuilding.

Reply:
cc @khushboo-rancher, she is planning negative tests, so we can base on that to write down all robot test cases w/o implementation first.

Reply:
Normally our test case will involve creating multiple volumes and dealing with different objects of different volumes, like: https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/tests/test_basic.py#L1584.

Is there a good way to write BDD style test step to deal with this kind of object referencing? like:
```
snapshot1 of volume1
backup2 of volume2
data1 in volume3
data2 in volume3
replica3 of volume3
node of replica3 of volume3 
... and so on
```

For the existing test cases, each test case is just a big function, and every object is just a local variable, so we don't need to worry about this. But for BDD, each test step (and its corresponding function) is independent, I don't have a clear idea about this yet. Can anyone shed some light on this?

Reply:
I think we can first pick some most complex test cases from the existing test code (like `test_listing_backup_volume`, `test_salvage_auto_crash_all_replicas`) or the negative test cases (like `Replica Rebuilding with replica deletion`) and try to translate them into robot framework test case definitions, then we know whether there is any obstacle.

Reply:
Take this test case definition for example:
> 
> ```
> *** Test Cases ***
> Replica Rebuilding With Replica Deletion
>     Given create Volume
>     and write random data to volume
>     When delete replica to trigger replica rebuilding
>     Then after replica rebuilt, the volume data is intact
>  
The original negative test case is:
```
Create a volume 5 GB with 3 replicas. r1, r2 and r3
Write in the volume.
Delete one replica to trigger the rebuilding. Let's say r1 is rebuilding.
While rebuilding is happening, delete other replica r2.
Wait for replica rebuilding r1 get finished. Delete replica r3.
Check the data
```
This requires delete different replica every time.

But for the step definition:
>     When delete replica to trigger replica rebuilding
It doesn't indicate which replica to be deleted, and if it doesn't store the state somewhere neither, probably it will always delete the same replica.

Reply:
How about below? Besides, you should be able to use some argument to make the below test run with different cases like different size fo volume or number of replica.

```robot
*** Test Cases ***
Replica Rebuilding With Replica Deletion
	Given create a volume 5 GB with 3 replicas # create 3 replicas with a convention name implicitly, like r1, r2, r3
	And write data to the volume
	
	When delete replica 1 to trigger the replica rebuilding # "replica 1" means r1, you should be able to save this variable (rebuilding_replica) via "Set Test Variable" builtin function provided by robot
	And during the rebuilding, delete replica 2 # you can check "rebuilding_replica" is rebuilding, then delete r2
	And wait for replica 1 rebuilt # just wait "rebuilding_replica" is ready
	And delete replica 3 # check r3
	
	Then check the data is intact
```